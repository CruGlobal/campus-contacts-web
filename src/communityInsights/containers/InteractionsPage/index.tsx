import React, { useContext } from 'react';
import { useTranslation } from 'react-i18next';
import moment from 'moment';

import Card from '../../components/Card';
import { AppContext } from '../../../appContext';
import StagesSummary from '../../components/StagesSummary';
import ImpactInfo from '../../components/ImpactInfo';
import FiltersChart from '../../components/FiltersChart';

import {
    INTERACTIONS_TOTAL_REPORT,
    INTERACTIONS_TOTAL_COMPLETED_REPORT,
    INTERACTIONS_COMPLETED_REPORT,
} from './queries';
// INTERACTIONS_TOTAL_REPORT INTERFACES
import { impactReportInteractionsCount } from './__generated__/impactReportInteractionsCount';
// INTERACTIONS_TOTAL_COMPLETED_REPORT INTERFACES
import {
    communityReportInteractions,
    communityReportInteractions_community_report_interactions as communityReportInteractionsEntry,
} from './__generated__/communityReportInteractions';
// INTERACTIONS_COMPLETED_REPORT INTERFACES
import {
    communityReportDaysInteractions,
    communityReportDaysInteractions_community_report_daysReport as communityReportDaysInteractionsRow,
    communityReportDaysInteractions_community_report_daysReport_interactionResults as communityReportDaysInteractionsStage,
} from './__generated__/communityReportDaysInteractions';

const InteractionsPage = () => {
    const { t } = useTranslation('insights');
    const { orgId } = useContext(AppContext);

    return (
        <div>
            <ImpactInfo
                query={INTERACTIONS_TOTAL_REPORT}
                text={(report: impactReportInteractionsCount) =>
                    t('interactions.taken', {
                        count: report.community.impactReport.interactionsCount,
                        people:
                            report.community.impactReport
                                .interactionsReceiversCount,
                    })
                }
                variables={{ id: orgId }}
            />
            <Card title={t('interactions.totalCompleted')}>
                <StagesSummary
                    query={INTERACTIONS_TOTAL_COMPLETED_REPORT}
                    variables={{ id: orgId }}
                    longNames={true}
                    mapData={(data: communityReportInteractions) =>
                        data.community.report.interactions.map(
                            (entry: communityReportInteractionsEntry) => ({
                                stage: entry.interactionType.name,
                                icon: entry.interactionType.name
                                    .toLowerCase()
                                    .split(' ')
                                    .join('-'),
                                count: entry.interactionCount,
                            }),
                        )
                    }
                />
            </Card>
            <Card
                title={t('interactions.completed')}
                subtitle={t('interactions.completedSubtitle')}
            >
                <FiltersChart
                    nullContent={'interactionsCompleted'}
                    query={INTERACTIONS_COMPLETED_REPORT}
                    variables={{ id: orgId }}
                    mapData={(data: communityReportDaysInteractions) =>
                        data.community.report.daysReport.map(
                            (row: communityReportDaysInteractionsRow) => ({
                                ['total']: row.interactionsCount,
                                ['stages']: row.interactionResults.map(
                                    (
                                        stage: communityReportDaysInteractionsStage,
                                    ) => ({
                                        name: stage.interactionType.name,
                                        count: stage.count,
                                    }),
                                ),
                                ['date']: row.date,
                            }),
                        )
                    }
                    currentDate={moment().toDate()}
                    label={t('interactions.legend')}
                />
            </Card>
        </div>
    );
};

export default InteractionsPage;
