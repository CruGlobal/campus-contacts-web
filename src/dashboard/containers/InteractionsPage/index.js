import React, { useContext } from 'react';
import { useTranslation } from 'react-i18next';
import moment from 'moment';
import _ from 'lodash';

import Card from '../../components/Card';
import AppContext from '../../appContext';
import StagesSummary from '../../components/StagesSummary';
import ImpactInfo from '../../components/ImpactInfo';
import FiltersChart from '../../components/FiltersChart';

import {
    INTERACTIONS_TOTAL_REPORT,
    INTERACTIONS_TOTAL_COMPLETED_REPORT,
    INTERACTIONS_COMPLETED_REPORT,
} from './queries';

const InteractionsPage = () => {
    const { t } = useTranslation('insights');
    const { orgId } = useContext(AppContext);

    return (
        <div>
            <ImpactInfo
                query={INTERACTIONS_TOTAL_REPORT}
                text={report =>
                    t('interactions.taken', {
                        count:
                            report.communityReport.impactReport
                                .interactionsCount,
                        people:
                            report.communityReport.impactReport
                                .interactionsReceiversCount,
                    })
                }
                variables={{ organizationId: orgId }}
            />
            <Card title={t('interactions.totalCompleted')}>
                <StagesSummary
                    query={INTERACTIONS_TOTAL_COMPLETED_REPORT}
                    variables={{
                        organizationId: orgId,
                    }}
                    mapData={data =>
                        data.communityReport.interactions.map(entry => ({
                            stage: entry.interactionType.name,
                            icon: entry.interactionType.name
                                .toLowerCase()
                                .split(' ')
                                .join('-'),
                            count: entry.interactionCount,
                        }))
                    }
                />
            </Card>
            <Card
                title={t('interactions.completed')}
                subtitle={t('interactions.completedSubtitle')}
            >
                <FiltersChart
                    query={INTERACTIONS_COMPLETED_REPORT}
                    mapData={data =>
                        data.communityReport.dayReport.map(row => ({
                            ['total']: row.interactionCount,
                            ['stages']: row.interactions.map(stage => ({
                                name: stage.interactionType.name,
                                count: stage.interactionCount,
                            })),
                            ['date']: row.date.substring(0, 2),
                        }))
                    }
                    countAverage={data =>
                        Math.floor(_.sumBy(data, 'total') / data.length)
                    }
                    label={t('interactions.legend')}
                />
            </Card>
        </div>
    );
};

export default InteractionsPage;

InteractionsPage.propTypes = {};
