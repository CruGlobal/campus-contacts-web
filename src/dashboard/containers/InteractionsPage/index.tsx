import React, { useContext } from 'react';
import { useTranslation } from 'react-i18next';
import moment from 'moment';

import Card from '../../components/Card';
import AppContext from '../../appContext';
import StagesSummary from '../../components/StagesSummary';
import ImpactInfo from '../../components/ImpactInfo';
import FiltersChart from '../../components/FiltersChart';

import {
    INTERACTIONS_TOTAL_REPORT,
    INTERACTIONS_TOTAL_COMPLETED_REPORT,
    INTERACTIONS_COMPLETED_REPORT,
} from './queries';

const InteractionsPage = () => {
    const { t } = useTranslation('insights');
    const { orgId }: any = useContext(AppContext);

    return (
        <div>
            <ImpactInfo
                query={INTERACTIONS_TOTAL_REPORT}
                text={report =>
                    t('interactions.taken', {
                        count: report.community.impactReport.interactionsCount,
                        people:
                            report.community.impactReport
                                .interactionsReceiversCount,
                    })
                }
                variables={{ id: orgId }}
            />
            <Card title={t('interactions.totalCompleted')}>
                <StagesSummary
                    query={INTERACTIONS_TOTAL_COMPLETED_REPORT}
                    variables={{ id: orgId }}
                    mapData={data =>
                        data.community.report.interactions.map(
                            (entry: any) => ({
                                stage: entry.interactionType.name,
                                icon: entry.interactionType.name
                                    .toLowerCase()
                                    .split(' ')
                                    .join('-'),
                                count: entry.interactionCount,
                            }),
                        )
                    }
                />
            </Card>
            <Card
                title={t('interactions.completed')}
                subtitle={t('interactions.completedSubtitle')}
            >
                <FiltersChart
                    query={INTERACTIONS_COMPLETED_REPORT}
                    variables={{ id: orgId }}
                    mapData={data =>
                        data.community.report.daysReport.map((row: any) => ({
                            ['total']: row.interactionsCount,
                            ['stages']: row.interactionResults.map(
                                (stage: any) => ({
                                    name: stage.interactionType.name,
                                    count: stage.count,
                                }),
                            ),
                            ['date']: row.date,
                        }))
                    }
                    currentDate={moment().toDate()}
                    label={t('interactions.legend')}
                />
            </Card>
        </div>
    );
};

export default InteractionsPage;

InteractionsPage.propTypes = {};
