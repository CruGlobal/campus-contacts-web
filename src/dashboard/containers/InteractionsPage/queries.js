import gql from 'graphql-tag';

const INTERACTIONS_TOTAL_REPORT = gql`
    query communityReport {
        communityReport {
            impactReport {
                interactionsCount
                interactionsReceiversCount
            }
        }
    }
`;

const INTERACTIONS_TOTAL_COMPLETED_REPORT = gql`
    query communityReport {
        communityReport {
            interactions {
                interactionCount
                interactionType {
                    name
                }
            }
        }
    }
`;

const INTERACTIONS_COMPLETED_REPORT = gql`
    query communityReport {
        communityReport {
            dayReport {
                date
                interactionCount
                interactions {
                    interactionCount
                    interactionType {
                        name
                    }
                }
            }
        }
    }
`;

export {
    INTERACTIONS_TOTAL_REPORT,
    INTERACTIONS_TOTAL_COMPLETED_REPORT,
    INTERACTIONS_COMPLETED_REPORT,
};
