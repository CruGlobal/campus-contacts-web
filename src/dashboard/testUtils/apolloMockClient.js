import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { addMockFunctionsToSchema } from 'graphql-tools';
import { buildClientSchema } from 'graphql/utilities';
import { createMockerFromIntrospection } from 'fraql/mock';
import gql from 'fraql';

import introspectionQuery from '../../../schema.json';

import { globalMocks } from './globalMocks';

export const createApolloMockClient = (mocks = {}) => {
    const schema = buildClientSchema(introspectionQuery);

    addMockFunctionsToSchema({
        schema,
        mocks: { ...globalMocks, ...mocks },
    });

    return new ApolloClient({
        cache: new InMemoryCache(),
        link: new SchemaLink({ schema }),
    });
};

export const mockFragment = (fragmentDocument, options) => {
    return createMockerFromIntrospection(introspectionQuery, {
        mocks: globalMocks,
    }).mockFragment(gql(fragmentDocument.loc.source.body), options);
};
