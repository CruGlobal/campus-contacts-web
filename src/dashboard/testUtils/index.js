import React, { ReactElement } from 'react';
import { ApolloProvider } from 'react-apollo-hooks';
import { render } from '@testing-library/react';

import { createApolloMockClient } from './apolloMockClient';
import defaultTheme from '../defaultTheme';
import { ThemeProvider } from 'emotion-theming';
import AppContext from '../appContext';

export const renderWithContext = (component, { mocks } = {}, context = {}) => {
    const mockApolloClient = createApolloMockClient(mocks);

    // Warning: don't call any functions in here that return new instances on every call.
    // All the props need to stay the same otherwise renderer won't work.
    const wrapper = ({ children }) => (
        <ApolloProvider client={mockApolloClient}>
            <AppContext.Provider value={context}>
                <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
            </AppContext.Provider>
        </ApolloProvider>
    );

    const renderResult = render(component, { wrapper });

    return {
        ...renderResult,
        snapshot: () => {
            expect(renderResult.container).toMatchSnapshot();
        },
    };
};
