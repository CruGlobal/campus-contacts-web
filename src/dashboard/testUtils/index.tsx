import React, { ReactNode } from 'react';
import { ApolloProvider } from 'react-apollo-hooks';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'emotion-theming';

import defaultTheme from '../defaultTheme';
import AppContext from '../appContext';

import { createApolloMockClient } from './apolloMockClient';

export const renderWithContext = (
    component: any,
    { mocks: mocks = {}, appContext = {} } = {},
) => {
    const mockApolloClient = createApolloMockClient(mocks);

    // Warning: don't call any functions in here that return new instances on every call.
    // All the props need to stay the same otherwise renderer won't work.
    const wrapper = ({ children }: { children: ReactNode }) => (
        <ApolloProvider client={mockApolloClient}>
            <AppContext.Provider value={appContext}>
                <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
            </AppContext.Provider>
        </ApolloProvider>
    );

    // @ts-ignore
    const renderResult = render(component, { wrapper });
    return {
        ...renderResult,
        snapshot: () => {
            expect(renderResult.container).toMatchSnapshot();
        },
    };
};
