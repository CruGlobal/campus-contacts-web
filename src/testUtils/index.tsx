import React, { ReactNode, ReactElement } from 'react';
import { ApolloProvider } from 'react-apollo-hooks';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'emotion-theming';
import { IMocks } from 'graphql-tools';

import defaultTheme from '../defaultTheme';
import {
    AppContext,
    AppContextValue,
    appContextDefaultValue,
} from '../appContext';

import { createApolloMockClient } from './apolloMockClient';

interface RenderWithContextParams {
    mocks?: IMocks;
    appContext?: AppContextValue;
}

export const renderWithContext = (
    component: ReactElement,
    {
        mocks: mocks = {},
        appContext = appContextDefaultValue,
    }: RenderWithContextParams = {},
) => {
    const mockApolloClient = createApolloMockClient(mocks);

    // Warning: don't call any functions in here that return new instances on every call.
    // All the props need to stay the same otherwise renderer won't work.
    const wrapper = ({ children }: { children?: ReactNode }) => (
        <ApolloProvider client={mockApolloClient}>
            <AppContext.Provider value={appContext}>
                <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
            </AppContext.Provider>
        </ApolloProvider>
    );

    const renderResult = render(component, { wrapper });
    return {
        ...renderResult,
        snapshot: () => {
            expect(renderResult.container).toMatchSnapshot();
        },
    };
};
