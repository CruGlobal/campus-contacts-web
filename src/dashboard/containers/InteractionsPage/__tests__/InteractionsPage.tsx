import React from 'react';

import { renderWithContext } from '../../../testUtils';
import InteractionsPage from '../';

describe('<PersonalStepsPage />', () => {
    it('should render properly in loading state', async () => {
        renderWithContext(<InteractionsPage />, {
            appContext: {
                orgId: 1,
            },
        }).snapshot();
    });

    // TODO: Add a test with mocking data when GraphQL schema contains above query
});
