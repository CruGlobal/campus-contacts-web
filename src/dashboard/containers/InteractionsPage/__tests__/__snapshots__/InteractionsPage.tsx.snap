// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PersonalStepsPage /> should render properly in loading state 1`] = `
<div>
  <div>
    <h1
      class="css-1c2npbj"
    >
      Loading...
    </h1>
    <div
      class="css-w2m1ij"
    >
      <h1
        class="css-5j0tpf"
      >
        Total Interactions Completed
      </h1>
      <h2
        class="css-140lofm"
      />
      <div
        class="css-rr4yq"
      >
        <div
          class="css-1ohf0ui"
        >
          Loading...
        </div>
      </div>
    </div>
    <div
      class="css-w2m1ij"
    >
      <h1
        class="css-5j0tpf"
      >
        Interactions Completed
      </h1>
      <h2
        class="css-140lofm"
      >
        Total interactions completed over time.
      </h2>
      <div
        class="css-rr4yq"
      >
        <div>
          Loading...
        </div>
      </div>
    </div>
  </div>
</div>
`;
